admin/services/data-access/headteacher-declaration.data.service.js

```diff
@@ -49,4 +49,98 @@ headteacherDeclarationDataService.sqlFindHdfForCheck = async (dfeNumber, checkWi
   return R.head(result)
 }
 
+/**
+ * Find count of pupils blocking hdf submission before check end date
+ * @param dfeNumber
+ * @return {Number}
+ */
+headteacherDeclarationDataService.sqlFindPupilsBlockingHdfBeforeCheckEndDate = async (dfeNumber) => {
+  const sql = `
+    SELECT COUNT(p.id) as pupilsCount
+    FROM [mtc_admin].[pupil] p
+    JOIN [mtc_admin].[school] s ON p.school_id = s.id
+    LEFT JOIN [mtc_admin].[pupilStatus] ps ON (p.pupilStatus_id = ps.id)
+    WHERE s.dfeNumber = @dfeNumber
+    AND ps.code NOT IN ('NOT_TAKING', 'COMPLETED')
+  `
+
+  const params = [
+    { name: 'dfeNumber', type: TYPES.Int, value: dfeNumber }
+  ]
+
+  const result = await sqlService.query(sql, params)
+  return R.path(['pupilsCount'], R.head(result))
+}
+
+/**
+ * Find count of pupils blocking hdf submission before check end date
+ * @param dfeNumber
+ * @return {Number}
+ */
+headteacherDeclarationDataService.sqlFindPupilsBlockingHdfAfterCheckEndDate = async (dfeNumber) => {
+  const sql = `
+    SELECT COUNT(p.id) as pupilsCount
+    FROM [mtc_admin].[pupil] p
+    JOIN [mtc_admin].[school] s ON p.school_id = s.id
+    LEFT JOIN [mtc_admin].[pupilStatus] ps ON (p.pupilStatus_id = ps.id)
+    LEFT JOIN (
+        SELECT *,
+            ROW_NUMBER() OVER (PARTITION BY pupil_id ORDER BY id DESC) as rank
+        FROM [mtc_admin].[check]
+        WHERE isLiveCheck = 1
+           ) lastCheck ON (lastCheck.pupil_id = p.id)
+    LEFT JOIN [mtc_admin].[checkStatus] cs ON (lastCheck.checkStatus_id = cs.id)
+    WHERE s.dfeNumber = @dfeNumber
+    AND (lastCheck.rank = 1 or lastCheck.rank IS NULL)
+    AND (ps.code != 'STARTED' OR cs.code != 'NTR')
+    AND ps.code NOT IN ('NOT_TAKING', 'COMPLETED') 
+  `
+
+  const params = [
+    { name: 'dfeNumber', type: TYPES.Int, value: dfeNumber }
+  ]
+
+  const result = await sqlService.query(sql, params)
+  return R.path(['pupilsCount'], R.head(result))
+}
+
+/**
+ * Fetch all pupils for a school by dfeNumber with their status codes and attendance reasons
+ * @param dfeNumber
+ * @returns {Promise<*>}
+ */
+headteacherDeclarationDataService.sqlFindPupilsWithStatusAndAttendanceReasons = async function (dfeNumber) {
+  const paramDfeNumber = { name: 'dfeNumber', type: TYPES.Int, value: dfeNumber }
+
+  const sql = `
+    SELECT
+       p.*,
```

Comment:
It would be better to pick up only those columns you need rather than '*'.

Reply:
@jon-shipley done